File: lessons/dsa/prefix-sums/examples/range_sums_test.cpp
#include <cstdlib>
#include <iostream>
#include <vector>

#include "range_sums.cpp"

namespace {
int failures = 0;
const std::vector<int> kValues = {3, 1, 4, 1, 5, 9};

void check(const char* name, bool passed) {
  if (!passed) {
    std::cout << "FAIL " << name << "\n";
    ++failures;
  }
}
}  // namespace

int main() {
  const std::vector<int> prefix = buildPrefix(kValues);
  const int size = static_cast<int>(kValues.size());

  for (int start = 0; start < size; ++start) {
    for (int end = start; end < size; ++end) {
      check("ranges agree",
            rangeTotalByPrefix(prefix, start, end) == rangeTotalByScan(kValues, start, end));
    }
  }

  check("the lesson range totals ten", rangeTotalByPrefix(prefix, 2, 4) == 10);
  check("a range starting at zero", rangeTotalByPrefix(prefix, 0, 0) == 3);
  check("the whole sequence", rangeTotalByPrefix(prefix, 0, size - 1) == 23);
  check("prefix is one longer", prefix.size() == kValues.size() + 1);
  check("an empty sequence", buildPrefix({}).size() == 1);
  check("negative values", rangeTotalByPrefix(buildPrefix({5, -3, 2}), 0, 2) == 4);

  if (failures > 0) {
    return EXIT_FAILURE;
  }
  std::cout << "All checks passed.\n";
  return EXIT_SUCCESS;
}

File: lessons/dsa/prefix-sums/examples/RangeSumsTest.java
import java.util.List;

public final class RangeSumsTest {
  private static int failures = 0;
  private static final List<Integer> VALUES = List.of(3, 1, 4, 1, 5, 9);

  private static void check(String name, boolean passed) {
    if (!passed) {
      System.out.println("FAIL " + name);
      failures++;
    }
  }

  public static void main(String[] args) {
    List<Integer> prefix = RangeSums.buildPrefix(VALUES);

    for (int start = 0; start < VALUES.size(); start++) {
      for (int end = start; end < VALUES.size(); end++) {
        check("ranges agree",
            RangeSums.rangeTotalByPrefix(prefix, start, end)
                == RangeSums.rangeTotalByScan(VALUES, start, end));
      }
    }

    check("the lesson range totals ten", RangeSums.rangeTotalByPrefix(prefix, 2, 4) == 10);
    check("a range starting at zero", RangeSums.rangeTotalByPrefix(prefix, 0, 0) == 3);
    check("the whole sequence",
        RangeSums.rangeTotalByPrefix(prefix, 0, VALUES.size() - 1) == 23);
    check("prefix is one longer", prefix.size() == VALUES.size() + 1);
    check("an empty sequence", RangeSums.buildPrefix(List.of()).equals(List.of(0)));
    check("negative values",
        RangeSums.rangeTotalByPrefix(RangeSums.buildPrefix(List.of(5, -3, 2)), 0, 2) == 4);

    if (failures > 0) {
      System.exit(1);
    }
    System.out.println("All checks passed.");
  }
}
